File: ch04/squareMatrix.py
#!/usr/bin/python

import sys

def main(argv):
    # TODO: figure out a way of deserializing matrix strings.
    pass

def squareMatrixMul(A, B):
    n = len(A)
    C = [[0]*n for x in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]
    return C

if __name__ == '__main__':
    main(sys.argv)

File: ch04/squareMatrix.c
#include <stdlib.h>
#include <stdio.h>

int *squareMatrixMul(int *A, int *B, int size);

// TODO: figure out how to deserialize a string into a matrix.
int main(int argc, char **argv)
{
    
    return 0;
}

int *squareMatrixMul(int *A, int *B, int size)
{
    int *C;

    // Allocate space for a square matrix.
    C = malloc(size * size * sizeof(*C));

    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < size; ++j) {
            for (int k = 0; k < size; ++k) {
                C[i*size + j] += A[i*size + k] * B[k*size + j];
            }
        }
    }

    return C;
}